File: analyzer/predict.py
from analyzer.fen_parser import fen_to_vector, vector_to_label, label_to_vector
from analyzer.network import forward_pass

def predict_model(arguments, network):
    data_file = arguments["data_file"]
    
    total = 0
    correct = 0
    
    with open(data_file, "r") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            
            parts = line.split()
            if len(parts) >= 7:
                # Format: FEN (6 parts) + label (1-2 words)
                fen = " ".join(parts[:6])
                expected = " ".join(parts[6:])
                has_expected = True
            else:
                fen = line
                has_expected = False
            
            try:
                input_vec = fen_to_vector(fen)
                output, _ = forward_pass(network, input_vec)
                prediction = vector_to_label(output)
                
                # Add color for Check/Checkmate (color indicates who is in check)
                if prediction in ["Check", "Checkmate"]:
                    parts = fen.split()
                    if len(parts) >= 2:
                        # The side to move is the one in check/checkmate
                        color = "White" if parts[1] == 'W' else "Black"
                        prediction = f"{prediction} {color}"
                
                if has_expected:
                    total += 1
                    is_correct = (prediction == expected)
                    if is_correct:
                        correct += 1
                    status = "✓" if is_correct else "✗"
                    print(f"{status} {prediction} (expected: {expected})")
                else:
                    print(prediction)
            except Exception as e:
                print(f"Error processing FEN: {e}")
    
    if total > 0:
        accuracy = (correct / total) * 100
        print(f"\n{'='*50}")
        print(f"Results: {correct}/{total} correct ({accuracy:.2f}%)")
        print(f"{'='*50}")

File: analyzer_cpp/predict.cpp
#include "predict.hpp"
#include "fen_parser.hpp"
#include "network.hpp"
#include <fstream>
#include <sstream>
#include <iostream>

void predict_model(const AnalyzerArgs& args, json::Value& network) {
    std::ifstream file(args.data_file);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open data file: " + args.data_file);
    }
    
    int total = 0;
    int correct = 0;
    std::string line;
    
    while (std::getline(file, line)) {
        if (line.empty()) continue;
        
        std::istringstream iss(line);
        std::vector<std::string> parts;
        std::string word;
        while (iss >> word) {
            parts.push_back(word);
        }
        
        std::string fen;
        std::string expected;
        bool has_expected = false;
        
        if (parts.size() >= 7) {
            fen = parts[0] + " " + parts[1] + " " + parts[2] + " " + 
                  parts[3] + " " + parts[4] + " " + parts[5];
            
            for (size_t i = 6; i < parts.size(); i++) {
                if (i > 6) expected += " ";
                expected += parts[i];
            }
            has_expected = true;
        } else {
            fen = line;
        }
        
        try {
            auto input = fen_to_vector(fen);
            ForwardCache cache;
            auto output = forward_pass(network, input, cache);
            std::string prediction = vector_to_label(output);
            
            // Add color for Check/Checkmate
            if (prediction == "Check" || prediction == "Checkmate") {
                std::istringstream fen_iss(fen);
                std::string board, turn;
                fen_iss >> board >> turn;
                if (!turn.empty()) {
                    std::string color = (turn == "W" || turn == "w") ? "White" : "Black";
                    prediction = prediction + " " + color;
                }
            }
            
            if (args.debug_mode && has_expected) {
                total++;
                bool is_correct = (prediction == expected);
                if (is_correct) correct++;
                
                std::string status = is_correct ? "✓" : "✗";
                std::cout << status << " " << prediction << " (expected: " << expected << ")" << std::endl;
            } else {
                std::cout << prediction << std::endl;
            }
        } catch (const std::exception& e) {
            std::cerr << "Error processing FEN: " << e.what() << std::endl;
        }
    }
    
    file.close();
    
    if (args.debug_mode && total > 0) {
        double accuracy = (double)correct / total * 100.0;
        std::cout << "\n==================================================\n";
        std::cout << "Results: " << correct << "/" << total << " correct (" << accuracy << "%)\n";
        std::cout << "==================================================\n";
    }
}
